docs/design-notes/qualifying-replicas-for-reads.md

```diff
@@ -0,0 +1,69 @@
+# Qualifying replicas for read operations
+
+## Intro
+
+When a read request arrives at the coordinator node, it is later
+forwarded to one or more replicas in order to read the actual data
+and return it back to the client. Each request has an associated
+timeout, which indicates how long the client is willing to wait
+for the query results.
+This timeout information can be leveraged to lift some pressure off
+a replica which is overloaded or dead.
+
+## Gathering last seen information
+
+The coordinator should be able to predict whether a replica is likely
+to respond to a request in time or not. This prediction should be as
+precise as possible, but at the same time have minimal overhead, because
+performing read requests is on the hot path. A simple estimation
+of how responsive a replica is likely to be is implemented as follows:
+ * each time a read request is sent to a replica, a timestamp is recorded
+ * each time a response is received from a replica, a timestamp is recorded
+ * the difference between these two timestamps is a cheap, rough estimate
+   of how long it may take this replica to respond
+Note that this estimation is not precise - it does not determine whether
+a `sent` timestamp and a `responded` timestamp belong to the same request,
+so these values may vary. However, assuming that the cluster is under
+continuous, uniform load, these values would get updated very often, and would
+hopefully approach the approximate round-trip time for a single request.
+Finally, when a replica was not contacted in a long while (e.g. 5s),
```

Comment:
I think this is the part that works.

Maybe we could estimate overload by looking at response time for past requests. So if we had a QUORUM read and replica A responded in 1ms and replica B responded in 5ms, that gives us some information about A and B. But it doesn't work for LOCAL_ONE queries.

Reply:
Yup, it would make more sense. It's actually also trivial to gather the response time for a single request. I'll take it into consideration for the next iteration, to actually make this estimate work in cases other than a killed node - in particular, a replica which constantly replies, but with a large lag.

Reply:
I think we should brainstorm this more. Maybe look at literature too.

Reply:
Certainly. I'll add an RFC prefix to this series to indicate that it's not really ready yet